compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp

```diff
@@ -113,7 +114,147 @@ static void specializeGenericTransposeOp(RewriterBase &rewriter,
 }
 
 //===----------------------------------------------------------------------===//
-// Transpose propagation
+// Other pattern helpers
+//===----------------------------------------------------------------------===//
+
+/// If the `op` is a ContractionOpInterface, return the generalized op. If the
+/// `op` is a linalg::GenericOp, then just return the generic op.
+static FailureOr<linalg::GenericOp>
+getGenericOpOrGeneralizeContraction(RewriterBase &rewriter, Operation *op) {
+  auto linalgOp = dyn_cast<linalg::LinalgOp>(op);
+  if (!linalgOp) {
+    return failure();
+  }
+  if (!isa<linalg::GenericOp>(linalgOp) &&
+      !linalg::isaContractionOpInterface(linalgOp)) {
+    return failure();
+  }
+  auto genericOp = dyn_cast<linalg::GenericOp>(op);
+  if (genericOp) {
+    return genericOp;
+  }
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPoint(linalgOp);
+  return linalg::generalizeNamedOp(rewriter, linalgOp);
+}
+
+//===----------------------------------------------------------------------===//
+// Transpose Bubbling Patterns
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+// Fuses a transpose with the init of a linalg.generic op or contraction op.
+// Contraction ops are generalized and then treated as a generic. For example,
+//
+// linalg.generic {
+//   indexing_maps = [affine_map<(d0, d1, d2) -> (d1, d0, d2)>,
+//                    affine_map<(d0, d1, d2) -> (d0, d1, d2)>]
+//   ins(%0 : tensor<2x7x5>) outs(%1 : tensor<7x2x5>)
+//
+// %2 = linalg.transpose ... permutation = [0, 2, 1] :
+//                           tensor<7x2x5> -> tensor<7x5x2>
+// Becomes
+//
+// linalg.generic {
```

Comment:
Is it possible to keep the outs indexing map still identity. That can cause issues later on.... Also this isnt bubbling the transpose up. It is kind of absorbing it in.... Questioning if we really need this. We can probably achieve this as a nemd op change? Can you give the real example that is driving this?

Reply:
The outs indexing map is kept as an identity. We do need a test for the actual case that motivated this change though. For context now, the motivating case is this:

```mlir
%3 = linalg.matmul ins(%0, %1 : tensor<8192x640xf16>, tensor<640x640xf16>)
       outs(%2 : tensor<8192x640xf16>) -> tensor<8192x640xf16>
%expanded = tensor.expand_shape %3 [[0, 1], [2, 3]]
       : tensor<8192x640xf16> into tensor<2x4096x10x64xf16>
%5 = linalg.transpose ins(%expanded : tensor<2x4096x10x64xf16>)
       outs(%4 : tensor<2x10x4096x64xf16>) permutation = [0, 2, 1, 3]
```